src/url_search_params.rs

```diff
@@ -0,0 +1,483 @@
+    /// ```
+    pub fn get_all(&self, key: &str) -> UrlSearchParamsEntry {
+        unsafe {
+            let strings = ffi::ada_search_params_get_all(self.0, key.as_ptr().cast(), key.len());
+            let size = ffi::ada_strings_size(strings);
+            UrlSearchParamsEntry::new(strings, size)
+        }
+    }
+
+    /// Returns all keys as an iterator
+    ///
+    /// ```
+    /// use ada_url::UrlSearchParams;
+    /// let params = UrlSearchParams::parse("a=1")
+    ///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+    /// let mut keys = params.keys();
+    /// assert!(keys.next().is_some());
+    pub fn keys(&self) -> UrlSearchParamsKeyIterator {
+        let iterator = unsafe { ffi::ada_search_params_get_keys(self.0) };
+        UrlSearchParamsKeyIterator::new(iterator)
+    }
+
+    /// Returns all keys as an iterator
+    ///
+    /// ```
+    /// use ada_url::UrlSearchParams;
+    /// let params = UrlSearchParams::parse("a=1")
+    ///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+    /// let mut values = params.values();
+    /// assert!(values.next().is_some());
+    pub fn values(&self) -> UrlSearchParamsValueIterator {
+        let iterator = unsafe { ffi::ada_search_params_get_values(self.0) };
+        UrlSearchParamsValueIterator::new(iterator)
+    }
+
+    /// Returns all entries as an iterator
+    ///
+    /// ```
+    /// use ada_url::UrlSearchParams;
+    /// let params = UrlSearchParams::parse("a=1")
+    ///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+    /// let mut entries = params.entries();
+    /// assert_eq!(entries.next(), Some(("a", "1")));
+    /// ```
+    pub fn entries(&self) -> UrlSearchParamsEntryIterator {
+        let iterator = unsafe { ffi::ada_search_params_get_entries(self.0) };
+        UrlSearchParamsEntryIterator::new(iterator)
+    }
+}
+
+impl core::hash::Hash for UrlSearchParams {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+#[cfg(feature = "std")]
+impl core::str::FromStr for UrlSearchParams {
+    type Err = ParseUrlError<Box<str>>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::parse(s).map_err(|ParseUrlError { input }| ParseUrlError {
+            input: input.into(),
+        })
+    }
+}
+
+/// Returns the stringified version of the UrlSearchParams struct.
+///
+/// ```
+/// use ada_url::UrlSearchParams;
+/// let params = UrlSearchParams::parse("a=1&b=2")
+///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+/// assert_eq!(params.to_string(), "a=1&b=2");
+/// ```
+impl core::fmt::Display for UrlSearchParams {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str(unsafe { ffi::ada_search_params_to_string(self.0).as_ref() })
+    }
+}
+
+#[cfg(feature = "std")]
+impl<Input> Extend<(Input, Input)> for UrlSearchParams
+where
+    Input: AsRef<str>,
```

Comment:
You could use two different generics here, that way you could use a `&str, Cow<'_, str>` or whatever other combination of key/value types, instead of being forced into using the same type for key/value.

Reply:
What's the syntax for supporting multiple generics for the same variable?